### PR TITLE
Revamped express server with file upload

### PR DESCRIPTION
**Pull request by Tr1ple-F**
**Features**
- Directory resource to view directory
- File upload with configurable parameters
- Configurable file name
- File serve
- Search function with RegEx search optional
- Configuration files (config.json, protocol.json, ui.json)

### CODE REVIEW

File: f3-explorer-server/upload.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            }
+
+            let dbo = db.db("f3expl");
+            let doc = {};
+        
+            protocol.params.forEach(element => {
+                if (element.type !== "boolean" && element.mapping === undefined) {
+                    doc[element.id] = req.body[element.id];
+                } else {
+                    doc[element.id] = (element.mapping.db_entry !== undefined) ? (req.body[element.id] === 'true' ? element.mapping.db_entry[0] : element.mapping.db_entry[1]) : (req.body[element.id] === 'true' ? true : false);
```
Comment (Tr1ple-F) [inline f3-explorer-server/upload.js:24]: substitution of non-ASCII chars does not work

File: .gitignore [hunk 2]
```diff
@@ -45,4 +45,15 @@ f3-explorer-server/yarn-error.log*
...
+# global config json
+f3-explorer-ui/config.json
+f3-explorer-server/config.json
+
+# global protocol json
+f3-explorer-ui/protocol.json
+f3-explorer-server/protocol.json
+
+# global ui json
+f3-explorer-ui/ui.json
```
Comment (Tr1ple-F) [inline .gitignore:59]: why don't we have new lines at file end

File: f3-explorer-server/search.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                // Internal Server Error
+                res.sendStatus(500);
+            }
+
+            let dbo = db.db("f3expl");
+            let qso = {};
+            protocol.params.forEach(element => {
+                if(req.query[element.id] !== undefined){
+                    if (element.type !== "boolean" && element.mapping === undefined) {
+                        if(req.query[element.id].charAt(0) === '~'){
```
Comment (Tr1ple-F) [inline f3-explorer-server/search.js:22]: Search should not be case-sensitive.

File: f3-explorer-server/main.js [hunk 4]
```diff
@@ -1,49 +1,99 @@
...
+const protocol = require('./protocol.json');
+const { filter, formatDate } = require('./helpers.js');
+const { uploadDB } = require('./upload.js');
+const { searchDB } = require('./search.js');
+
+// Initialize multer storage
+const storage = multer.diskStorage({
+  destination: function(req, file, cb) {
+    // Save in public directory
+    cb(null, 'public/');
```
Comment (XSpielinbox) [inline f3-explorer-server/main.js:22]: What about saving in external directory?

File: protocol.json [hunk 5]
```diff
@@ -0,0 +1,59 @@
+{
+    "filename": "%schuljahr%_%jahrgang%_%fach%_%kursart%_%lehrer-kurz%_%lehrer-vname%_%lehrer-nname%%isVergleich%_%timestamp%.pdf",
+    "allowed_extensions": [
```
Comment (XSpielinbox) [inline protocol.json:3]: In general ok; many corrections needed

File: ui.json [hunk 6]
```diff
@@ -0,0 +1,117 @@
+{
+    "ui-elements": [
+        {
+            "target": "schuljahr",
```
Comment (XSpielinbox) [inline ui.json:4]: In general ok; many corrections needed

File: f3-explorer-server/main.js [hunk 7]
```diff
@@ -1,49 +1,99 @@
...
+
+  filename: function(req, file, cb) {
+    // Map path
+    let path = protocol.filename;
+    protocol.params.forEach(element => {
+        if (element.type !== "boolean" && element.mapping === undefined) {
+          let cleaned = req.body[element.id].replace(new RegExp(' ', 'gm'), '--').replace(new RegExp('%', 'gm'), '-');
+          path = path.replace(new RegExp('%' + element.id + '%', 'gm'), cleaned);
+        } else {
+            path = path.replace(new RegExp('%' + element.id + '%', 'gm'), (element.mapping.filename !== undefined) ? (req.body[element.id] === 'true' ? element.mapping.filename[0] : element.mapping.filename[1]) : (req.body[element.id] === 'true' ? true : false));
```
Comment (XSpielinbox) [inline f3-explorer-server/main.js:33]: Not readable

File: config.json [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{
+    "server_port": 7777,
+    "frontend_port": 5000,
```
Comment (XSpielinbox) [inline config.json:3]: Run via http for beeing able to omit port?

File: f3-explorer-server/main.js [hunk 9]
```diff
@@ -1,49 +1,99 @@
...
+        }
+    });
+    path = path.replace(new RegExp('%timestamp%', 'gm'), formatDate('YmdHis'));
+    cb(null, path);
+  }
+});
+
+// Start express server
 const app = express();
-const port = 7777;
```
Comment (XSpielinbox) [inline f3-explorer-server/main.js:6]: Why duplicate assignment?

File: f3-explorer-server/package.json [hunk 10]
```diff
@@ -4,8 +4,7 @@
   "description": "A REST-API for a Raspi sensor",
```
Comment (XSpielinbox) [inline f3-explorer-server/package.json:4]: Outdated

File: f3-explorer-server/package.json [hunk 11]
```diff
@@ -4,8 +4,7 @@
...
   "main": "main.js",
   "scripts": {
-    "start": "node main.js",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "start": "copyfiles ../protocol.json ./protocol.json && copyfiles ../config.json ./config.json && node main.js"
```
Comment (XSpielinbox) [inline f3-explorer-server/package.json:7]: Why is no copy of ui.json needed?

File: f3-explorer-server/search.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+                            qso[element.id] = new RegExp(req.query[element.id].substring(1));
+                        } else {
+                            qso[element.id] = req.query[element.id];
+                        }
+                    } else {
+                        qso[element.id] = (element.mapping.db_entry !== undefined) ? (req.query[element.id] === 'true' ? element.mapping.db_entry[0] : element.mapping.db_entry[1]) : (req.query[element.id] === 'true' ? true : false);
```
Comment (XSpielinbox) [inline f3-explorer-server/search.js:28]: Not readable

File: f3-explorer-server/upload.js (hunk 1 above)
Comment (XSpielinbox) [reply f3-explorer-server/upload.js:24]: Should be done at frontend. Line is not readable

File: f3-explorer-server/main.js (hunk 4 above)
Comment (Tr1ple-F) [reply f3-explorer-server/main.js:22]: Is implemented

File: f3-explorer-server/search.js (hunk 3 above)
Comment (Tr1ple-F) [reply f3-explorer-server/search.js:22]: Should that be accounted for at the frontend?

File: f3-explorer-server/main.js (hunk 9 above)
Comment (Tr1ple-F) [reply f3-explorer-server/main.js:6]: Fixed

File: f3-explorer-server/package.json (hunk 10 above)
Comment (Tr1ple-F) [reply f3-explorer-server/package.json:4]: Fixed

File: f3-explorer-server/package.json (hunk 11 above)
Comment (Tr1ple-F) [reply f3-explorer-server/package.json:7]: Because ui.json is only used by the frontend

File: f3-explorer-server/main.js (hunk 7 above)
Comment (Tr1ple-F) [reply f3-explorer-server/main.js:33]: Fixed with multiline code

File: f3-explorer-server/search.js (hunk 12 above)
Comment (Tr1ple-F) [reply f3-explorer-server/search.js:28]: Fixed with multiline code

File: f3-explorer-server/package.json [hunk 13]
```diff
@@ -16,9 +15,12 @@
   "author": "Finlay Fehlauer",
   "license": "MIT",
```
Comment (XSpielinbox) [inline f3-explorer-server/package.json:16]: Outdated licence

**Review by XSpielinbox — APPROVED** [review]

**Pull request merged**